pkg/webhook/pod/pd_deleter.go

```diff
@@ -0,0 +1,175 @@
+	}
+
+	// NotInOrdinal represents this is an scale-in operation, we need to delete this member in pd cluster
+	if !isInOrdinal {
+		return pc.admitDeleteExceedReplicasPDPod(pod, tcName, pdClient)
+	}
+
+	// If there is an pd pod deleting operation during upgrading, we should
+	// check the pd pod which had been upgraded before were all health
+	if isUpgrading {
+		klog.Infof("receive delete pd pod[%s/%s] of tc[%s/%s] is upgrading, make sure former pd upgraded status was health", namespace, name, namespace, tcName)
+		err = checkFormerPDPodStatus(pc.kubeCli, pdClient, tc, namespace, ordinal, *ownerStatefulSet.Spec.Replicas)
+		if err != nil {
+			return util.ARFail(err)
+		}
+	}
+
+	leader, err := pdClient.GetPDLeader()
+	if err != nil {
+		klog.Errorf("tc[%s/%s] fail to get pd leader %v,refuse to delete pod[%s/%s]", namespace, tc.Name, err, namespace, name)
+		return util.ARFail(err)
+	}
+
+	klog.Infof("tc[%s/%s]'s pd leader is pod[%s/%s] during deleting pod[%s/%s]", namespace, tc.Name, namespace, leader.Name, namespace, name)
+	if leader.Name == name {
+		return pc.admitDeletePDLeader(pod, tc, pdClient, ordinal)
+	}
+
+	klog.Infof("pod[%s/%s] is not pd-leader,admit to delete", namespace, name)
+	return util.ARSuccess()
+}
+
+// this pod is not a member of pd cluster currently, it could be deleted from pd cluster already or haven't registered in pd cluster
+// we need to check whether this pd pod has been ensured wouldn't be a member in pd cluster
+func (pc *PodAdmissionControl) admitDeleteNonPDMemberPod(IsDeferDeleting, isInOrdinal bool, pod *corev1.Pod, ownerStatefulSet *apps.StatefulSet, tc *v1alpha1.TidbCluster, pdClient pdapi.PDClient, ordinal int32) *v1beta1.AdmissionResponse {
+
+	name := pod.Name
+	namespace := pod.Namespace
+	tcName := tc.Name
+
+	// check whether this pod has been ensured wouldn't be a member in pd cluster
+	if IsDeferDeleting {
+		// when pd scale in, we should delete member first and finally edit its pvc and admit to delete pod.
+		// If we edit pvc first and finally delete pd member during scale in,
+		// it can be error that pd scale in From 4 to 3,the pvc was edited successfully and
+		// the pd member was fail to deleted. If the pd scale was recover to 4 at that time,
+		// it would be existed an pd-3 instance with its deferDeleting label Annotations PVC.
+		// And the pvc can be deleted during upgrading if we use create pod webhook in future.
+		if !isInOrdinal {
+			err := addDeferDeletingToPVC(pc, tc, ownerStatefulSet.Name, namespace, ordinal)
+			if err != nil {
+				klog.Infof("tc[%s/%s]'s pod[%s/%s] failed to update pvc,%v", namespace, tcName, namespace, name, err)
+				return util.ARFail(err)
+			}
+		}
+		klog.Infof("pd pod[%s/%s] is not member of tc[%s/%s],admit to delete", namespace, name, namespace, tcName)
+		return util.ARSuccess()
+
+	}
+	err := pdClient.DeleteMember(name)
+	if err != nil {
+		return util.ARFail(err)
+	}
+	err = addDeferDeletingToPDPod(pc, pod)
```

Comment:
Many function names include `PD`, maybe we can make them more general if some of them can be reused when supporting TiKV & TiDB.

Reply:
I think it could be done during tikv delete pod handler. It's hard to say whether those func could be a common method currently.

Reply:
The defer deleting annotation cannot resolve the case where the PD is not a member yet here but becomes a member later, but the Pod will be removed in the next request.